File: system-test/java/src/main/java/com/optivem/shop/testkit/driver/adapter/external/tax/BaseTaxDriver.java
package com.optivem.shop.testkit.driver.adapter.external.tax;

import com.optivem.shop.testkit.driver.port.external.tax.TaxDriver;

import com.optivem.shop.testkit.driver.adapter.external.tax.client.BaseTaxClient;
import com.optivem.shop.testkit.driver.port.external.tax.dtos.GetTaxResponse;
import com.optivem.shop.testkit.driver.port.shared.dtos.ErrorResponse;
import com.optivem.shop.testkit.common.Closer;
import com.optivem.shop.testkit.common.Result;

public abstract class BaseTaxDriver<TClient extends BaseTaxClient> implements TaxDriver {
    protected final TClient client;

    protected BaseTaxDriver(TClient client) {
        this.client = client;
    }

    @Override
    public void close() {
        Closer.close(client);
    }

    @Override
    public Result<Void, ErrorResponse> goToTax() {
        return client.checkHealth()
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }

    @Override
    public Result<GetTaxResponse, ErrorResponse> getTaxRate(String country) {
        return client.getCountry(country)
                .map(taxRateResponse -> GetTaxResponse.builder()
                        .country(taxRateResponse.getId())
                        .taxRate(taxRateResponse.getTaxRate())
                        .build())
                .mapError(ext -> ErrorResponse.builder().message(ext.getMessage()).build());
    }
}

File: system-test/typescript/src/testkit/driver/adapter/external/tax/BaseTaxDriver.ts
import type { Result } from '../../../../common/result.js';
import type { TaxErrorResponse } from '../../../port/external/tax/dtos/TaxErrorResponse.js';
import type { GetTaxResponse } from '../../../port/external/tax/dtos/GetTaxResponse.js';
import type { ReturnsTaxRateRequest } from '../../../port/external/tax/dtos/ReturnsTaxRateRequest.js';
import type { TaxDriver } from '../../../port/external/tax/tax-driver.js';
import type { BaseTaxClient } from './client/BaseTaxClient.js';

export abstract class BaseTaxDriver<TClient extends BaseTaxClient> implements TaxDriver {
  protected constructor(protected readonly client: TClient) {}

  async goToTax(): Promise<Result<void, TaxErrorResponse>> {
    return this.client.checkHealth();
  }

  async getTaxRate(country: string): Promise<Result<GetTaxResponse, TaxErrorResponse>> {
    return this.client.getTaxRate(country);
  }

  abstract returnsTaxRate(request: ReturnsTaxRateRequest): Promise<Result<void, TaxErrorResponse>>;
  abstract close(): Promise<void>;
}
